enhancements/sig-policy/122-expanded-hub-templates-access/README.md

```diff
@@ -0,0 +1,310 @@
+Governance Policy Propagator, and instead, it will rely on the Kubernetes API performing authorization checks of the
+service account. Note that due to the event-driven nature of hub templates, the service account must have `list` and
+`watch` permissions for the objects that the hub templates reference.
+
+Note that for backwards compatibility, if the namespace is not specified in a hub template object lookup, the root
+policy namespace will be the default, however, the service account must still have the appropriate permissions in the
+root policy namespace to perform the lookup.
+
+See an example of such a policy below:
+
+```yaml
+apiVersion: policy.open-cluster-management.io/v1
+kind: Policy
+metadata:
+  name: test-config
+  namespace: test-policies
+spec:
+  disabled: false
+  remediationAction: enforce
+  hubTemplateOptions:
+    serviceAccountName: testsa
+  policy-templates:
+    - objectDefinition:
+        apiVersion: policy.open-cluster-management.io/v1
+        kind: ConfigurationPolicy
+        metadata:
+          name: test-config
+        spec:
+          remediationAction: enforce
+          severity: high
+          object-templates:
+            - complianceType: musthave
+              objectDefinition:
+                apiVersion: v1
+                data:
+                  field1: '{{hub fromConfigMap "some-other-namespace" "test-configmap" "test-field1" hub}}'
+                  field2: '{{hub fromConfigMap "some-other-namespace" "test-configmap" "test-field1" hub}}'
+                kind: ConfigMap
+                metadata:
+                  name: my-configmap
+                  namespace: test
+```
+
+### User Stories
+
+#### Story 1
+
+As a policy user, I would like to specify hub templates that reference objects outside of the root policy namespace so
+that I can stop defining additional policies on `local-cluster` to sync the required information to the root policy
+namespace.
+
+### Implementation Details/Notes/Constraints
+
+Although the user experience is rather simple, there's complexity in the implementation. One aspect is that the number
+of service accounts is not known at startup and can constantly change. It is also a bit complex to prevent duplication
+of Kubernetes API watches, underlying TCP connections, and tokens if multiple policies reference the same service
+account.
+
+Currently, the Governance Policy Propagator only uses a single `TemplateResolver` client from the
+[go-template-utils](https://github.com/stolostron/go-template-utils) library and sets restrictions based on the root
+policy namespace. That will still remain the default, however, we'll need exactly one `TemplateResolver` client per
+service account. To do this, I recommend having a new `sync.Map` field named `TemplateResolvers` on the
+`ReplicatedPolicyReconciler` struct and each key would be a `NamespacedName` from `k8s.io/apimachinery/pkg/types`
```

Comment:
I think we have used this pattern of a `sync.Map` to things with individual mutexes before, but it feels a little bit clunky to me. I wonder if any other teams have patterns they use for things like this, just to compare to.

Reply:
We do use this pattern and I agree it's a bit clunky but the `sync.Map` is required due to the dynamic nature of the keys since they can change at runtime and can come and go. The alternative is a global lock on a normal map but `sync.Map` is optimized for multiple reads and few writes:
https://pkg.go.dev/sync#Map

One thing we could try is [sync.OnceValue](https://pkg.go.dev/sync#OnceValue) instead of a lock in the struct for `sync.Map` but it uses locks under the hood so we may be better off managing the lock ourselves for efficiency.